### DISCUSSION TITLE
Apparent deadlock setting optional Shared property to nil

### DISCUSSION THREAD

**Discussion by humblehacker** [pointfreeco/swift-composable-architecture#3175] [Q&A] [answered]
Please see: https://github.com/humblehacker/TCASharedState2

In this example, I have a reducer with two children. The first has an optional shared property, while the second has a non-optional. Tapping the "Toggle" button should toggle the optional property between `0` and `nil`, but when it attempts to set the property to `nil`, the app freezes. 

Oddly, this only happens if there is at least one other shared property in the app. To see this, comment out `Child2View` from the body of `ExampleView`, and try again -- the app no longer freezes.

So, my question is, what am I doing wrong?

Note that this and my previous [bug report turned discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions/3171) were discovered while attempting to isolate the source of a problem encountered while using shared state in my app.

**Comment by mbrandonw** [comment, marked as answer]
Hi @humblehacker, thanks for the report. Yeah this one is a real issue, and we should have made `withLock` run on the `@MainActor` from the beginning. We thought about making `@Shared` main actor entirely, but that seemed to be going too far. However, `withLock` _definitely_ needs to be `@MainActor`.

I have a fix PR'd [here](https://github.com/pointfreeco/swift-composable-architecture/pull/3178) and it will be released with 1.12.
